File: frontend/iosApp/Memories/Sources/Memories/DI/AuthenticatedContainer.swift
import Foundation
import SwiftData
import APIGateways
import Domains
import Repositories
import UseCases
@preconcurrency import Shared

@MainActor
public final class AuthenticatedContainer {
    private let token: String
    private let userId: Int

    public init(token: String, userId: Int) {
        self.token = token
        self.userId = userId
    }

    // MARK: - Infrastructure

    private lazy var database: SwiftDatabase = {
        try! SwiftDatabase.create(
            for: userId,
            modelTypes: [
                LocalAlbum.self,
                LocalUser.self,
                LocalMemory.self,
                LocalSyncOperation.self
            ]
        )
    }()

    // MARK: - KMP API Client

    private lazy var kmpApiClient: Shared.AuthenticatedApiClient = {
        Shared.AuthenticatedApiClient(baseUrl: AppConfig.baseURL.absoluteString, apiToken: token)
    }()

    // MARK: - KMP Gateways

    private lazy var kmpUserGateway: Shared.UserGateway = {
        Shared.UserGatewayImpl(apiClient: kmpApiClient)
    }()

    private lazy var kmpAlbumGateway: Shared.AlbumGateway = {
        Shared.AlbumGatewayImpl(apiClient: kmpApiClient)
    }()

    private lazy var kmpMemoryGateway: Shared.MemoryGateway = {
        Shared.MemoryGatewayImpl(apiClient: kmpApiClient)
    }()

    // MARK: - Swift Repositories (SwiftData)

    private lazy var swiftUserRepository: Repositories.UserRepository = {
        Repositories.UserRepository(userId: userId, database: database)
    }()

    private lazy var swiftAlbumRepository: Repositories.AlbumRepository = {
        Repositories.AlbumRepository(database: database)
    }()

    private lazy var swiftMemoryRepository: Repositories.MemoryRepository = {
        Repositories.MemoryRepository(database: database)
    }()

    private lazy var swiftSyncQueueRepository: Repositories.SyncQueueRepository = {
        Repositories.SyncQueueRepository(database: database)
    }()

    private lazy var swiftImageStorageRepository: Repositories.ImageStorageRepository = {
        Repositories.ImageStorageRepository(userId: userId)
    }()

    public var reachabilityRepository: ReachabilityRepositoryProtocol {
        AppConfig.reachabilityRepository
    }

    // MARK: - KMP Repository Bridges (Swift -> KMP)

    private lazy var userRepositoryBridge: UserRepositoryBridgeImpl = {
        UserRepositoryBridgeImpl(repository: swiftUserRepository)
    }()

    private lazy var albumRepositoryBridge: AlbumRepositoryBridgeImpl = {
        AlbumRepositoryBridgeImpl(repository: swiftAlbumRepository)
    }()

    private lazy var memoryRepositoryBridge: MemoryRepositoryBridgeImpl = {
        MemoryRepositoryBridgeImpl(repository: swiftMemoryRepository)
    }()

    private lazy var syncQueueRepositoryBridge: SyncQueueRepositoryBridgeImpl = {
        SyncQueueRepositoryBridgeImpl(repository: swiftSyncQueueRepository)
    }()

    private lazy var imageStorageRepositoryBridge: ImageStorageRepositoryBridgeImpl = {
        ImageStorageRepositoryBridgeImpl(repository: swiftImageStorageRepository)
    }()

    private lazy var reachabilityRepositoryBridge: ReachabilityRepositoryBridgeImpl = {
        ReachabilityRepositoryBridgeImpl(repository: reachabilityRepository)
    }()

    private lazy var authSessionRepositoryBridge: AuthSessionRepositoryBridgeImpl = {
        AuthSessionRepositoryBridgeImpl(repository: AppConfig.authSessionRepository)
    }()

    // MARK: - KMP Repositories (via Bridge)

    private lazy var kmpUserRepository: Shared.UserRepository = {
        Shared.UserRepositoryImpl(bridge: userRepositoryBridge)
    }()

    private lazy var kmpAlbumRepository: Shared.AlbumRepository = {
        Shared.AlbumRepositoryImpl(bridge: albumRepositoryBridge)
    }()

    private lazy var kmpMemoryRepository: Shared.MemoryRepository = {
        Shared.MemoryRepositoryImpl(bridge: memoryRepositoryBridge)
    }()

    private lazy var kmpSyncQueueRepository: Shared.SyncQueueRepository = {
        Shared.SyncQueueRepositoryImpl(bridge: syncQueueRepositoryBridge)
    }()

    private lazy var kmpImageStorageRepository: Shared.ImageStorageRepository = {
        Shared.ImageStorageRepositoryImpl(bridge: imageStorageRepositoryBridge)
    }()

    private lazy var kmpReachabilityRepository: Shared.ReachabilityRepository = {
        Shared.ReachabilityRepositoryImpl(bridge: reachabilityRepositoryBridge)
    }()

    private lazy var kmpAuthSessionRepository: Shared.AuthSessionRepository = {
        Shared.AuthSessionRepositoryImpl(bridge: authSessionRepositoryBridge)
    }()

    // MARK: - KMP SyncQueueService

    private lazy var kmpSyncQueueService: Shared.SyncQueueService = {
        Shared.SyncQueueServiceImpl(
            syncQueueRepository: kmpSyncQueueRepository,
            albumRepository: kmpAlbumRepository,
            memoryRepository: kmpMemoryRepository,
            userRepository: kmpUserRepository,
            albumGateway: kmpAlbumGateway,
            memoryGateway: kmpMemoryGateway,
            userGateway: kmpUserGateway,
            imageStorageRepository: kmpImageStorageRepository,
            reachabilityRepository: kmpReachabilityRepository
        )
    }()

    // MARK: - KMP UseCases

    public lazy var splashUseCase: SplashUseCaseProtocol = {
        let kmpUseCase = Shared.SplashUseCaseImpl(
            userGateway: kmpUserGateway,
            userRepository: kmpUserRepository,
            authSessionRepository: kmpAuthSessionRepository,
            reachabilityRepository: kmpReachabilityRepository,
            syncQueueRepository: kmpSyncQueueRepository
        )
        return SplashUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

    public lazy var albumListUseCase: AlbumListUseCaseProtocol = {
        let kmpUseCase = Shared.AlbumListUseCaseImpl(
            userRepository: kmpUserRepository,
            albumRepository: kmpAlbumRepository,
            albumGateway: kmpAlbumGateway,
            reachabilityRepository: kmpReachabilityRepository,
            syncQueueService: kmpSyncQueueService,
            syncQueueRepository: kmpSyncQueueRepository
        )
        return AlbumListUseCaseAdapter(
            kmpUseCase: kmpUseCase,
            reachabilityRepository: reachabilityRepository
        )
    }()

    public lazy var userProfileUseCase: UserProfileUseCaseProtocol = {
        let kmpUseCase = Shared.UserProfileUseCaseImpl(
            userGateway: kmpUserGateway,
            userRepository: kmpUserRepository,
            authSessionRepository: kmpAuthSessionRepository,
            syncQueueService: kmpSyncQueueService,
            reachabilityRepository: kmpReachabilityRepository,
            imageStorageRepository: kmpImageStorageRepository
        )
        return UserProfileUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

    public lazy var albumFormUseCase: AlbumFormUseCaseProtocol = {
        let kmpUseCase = Shared.AlbumFormUseCaseImpl(
            albumRepository: kmpAlbumRepository,
            albumGateway: kmpAlbumGateway,
            syncQueueService: kmpSyncQueueService,
            reachabilityRepository: kmpReachabilityRepository,
            imageStorageRepository: kmpImageStorageRepository
        )
        return AlbumFormUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

    public lazy var memoryFormUseCase: MemoryFormUseCaseProtocol = {
        let kmpUseCase = Shared.MemoryFormUseCaseImpl(
            memoryRepository: kmpMemoryRepository,
            memoryGateway: kmpMemoryGateway,
            syncQueueService: kmpSyncQueueService,
            reachabilityRepository: kmpReachabilityRepository,
            imageStorageRepository: kmpImageStorageRepository
        )
        return MemoryFormUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

    public lazy var albumDetailUseCase: AlbumDetailUseCaseProtocol = {
        let kmpUseCase = Shared.AlbumDetailUseCaseImpl(
            memoryRepository: kmpMemoryRepository,
            albumRepository: kmpAlbumRepository,
            albumGateway: kmpAlbumGateway,
            memoryGateway: kmpMemoryGateway,
            reachabilityRepository: kmpReachabilityRepository
        )
        return AlbumDetailUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

    public lazy var syncQueuesUseCase: SyncQueuesUseCaseProtocol = {
        let kmpUseCase = Shared.SyncQueuesUseCaseImpl(
            syncQueueRepository: kmpSyncQueueRepository,
            albumRepository: kmpAlbumRepository,
            memoryRepository: kmpMemoryRepository,
            userRepository: kmpUserRepository
        )
        return SyncQueuesUseCaseAdapter(kmpUseCase: kmpUseCase)
    }()

}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/di/AuthenticatedContainer.kt
package com.example.memoriesapp.di

import com.example.memoriesapp.api.client.AuthenticatedApiClient
import com.example.memoriesapp.data.repository.AlbumRepositoryImpl
import com.example.memoriesapp.data.repository.ImageStorageRepositoryImpl
import com.example.memoriesapp.data.repository.MemoryRepositoryImpl
import com.example.memoriesapp.data.repository.SyncQueueRepositoryImpl
import com.example.memoriesapp.data.repository.UserRepositoryImpl
import com.example.memoriesapp.gateway.impl.AlbumGatewayImpl
import com.example.memoriesapp.gateway.impl.MemoryGatewayImpl
import com.example.memoriesapp.gateway.impl.UserGatewayImpl
import com.example.memoriesapp.repository.AuthSessionRepository
import com.example.memoriesapp.repository.ReachabilityRepository
import com.example.memoriesapp.usecase.AlbumDetailUseCase
import com.example.memoriesapp.usecase.AlbumFormUseCase
import com.example.memoriesapp.usecase.AlbumListUseCaseWrapper
import com.example.memoriesapp.usecase.MemoryFormUseCase
import com.example.memoriesapp.usecase.SplashUseCase
import com.example.memoriesapp.usecase.SyncQueuesUseCase
import com.example.memoriesapp.usecase.UserProfileUseCase
import com.example.memoriesapp.usecase.impl.AlbumDetailUseCaseImpl
import com.example.memoriesapp.usecase.impl.AlbumFormUseCaseImpl
import com.example.memoriesapp.usecase.impl.AlbumListUseCaseImpl
import com.example.memoriesapp.usecase.impl.MemoryFormUseCaseImpl
import com.example.memoriesapp.usecase.impl.SplashUseCaseImpl
import com.example.memoriesapp.usecase.impl.SyncQueueServiceImpl
import com.example.memoriesapp.usecase.impl.SyncQueuesUseCaseImpl
import com.example.memoriesapp.usecase.impl.UserProfileUseCaseImpl

/**
 * DI container for authenticated user session.
 * Contains all use cases that require authentication.
 */
class AuthenticatedContainer(
    private val context: android.content.Context,
    private val token: String,
    private val userId: Int,
    private val baseUrl: String,
    private val authSessionRepository: AuthSessionRepository,
    private val reachabilityRepository: ReachabilityRepository
) {
    // Repositories (in-memory, no persistence)
    private val userRepository = UserRepositoryImpl(userId)
    private val albumRepository = AlbumRepositoryImpl()
    private val memoryRepository = MemoryRepositoryImpl()
    private val syncQueueRepository = SyncQueueRepositoryImpl()
    private val imageStorageRepository = ImageStorageRepositoryImpl(context)

    // API Client (authenticated)
    private val apiClient = AuthenticatedApiClient(baseUrl, token)

    // Gateways
    private val userGateway = UserGatewayImpl(apiClient)
    private val albumGateway = AlbumGatewayImpl(apiClient)
    private val memoryGateway = MemoryGatewayImpl(apiClient)

    // SyncQueueService
    private val syncQueueService = SyncQueueServiceImpl(
        syncQueueRepository = syncQueueRepository,
        albumRepository = albumRepository,
        memoryRepository = memoryRepository,
        userRepository = userRepository,
        albumGateway = albumGateway,
        memoryGateway = memoryGateway,
        userGateway = userGateway,
        imageStorageRepository = imageStorageRepository,
        reachabilityRepository = reachabilityRepository
    )

    // UseCases
    val splashUseCase: SplashUseCase by lazy {
        SplashUseCaseImpl(
            userGateway = userGateway,
            userRepository = userRepository,
            authSessionRepository = authSessionRepository,
            reachabilityRepository = reachabilityRepository,
            syncQueueRepository = syncQueueRepository
        )
    }

    val albumListUseCase: AlbumListUseCaseWrapper by lazy {
        val impl = AlbumListUseCaseImpl(
            userRepository = userRepository,
            albumRepository = albumRepository,
            albumGateway = albumGateway,
            reachabilityRepository = reachabilityRepository,
            syncQueueService = syncQueueService,
            syncQueueRepository = syncQueueRepository
        )
        AlbumListUseCaseWrapper(impl, reachabilityRepository)
    }

    val albumDetailUseCase: AlbumDetailUseCase by lazy {
        AlbumDetailUseCaseImpl(
            memoryRepository = memoryRepository,
            albumRepository = albumRepository,
            albumGateway = albumGateway,
            memoryGateway = memoryGateway,
            reachabilityRepository = reachabilityRepository
        )
    }

    val albumFormUseCase: AlbumFormUseCase by lazy {
        AlbumFormUseCaseImpl(
            albumRepository = albumRepository,
            albumGateway = albumGateway,
            syncQueueService = syncQueueService,
            reachabilityRepository = reachabilityRepository,
            imageStorageRepository = imageStorageRepository
        )
    }

    val memoryFormUseCase: MemoryFormUseCase by lazy {
        MemoryFormUseCaseImpl(
            memoryRepository = memoryRepository,
            memoryGateway = memoryGateway,
            syncQueueService = syncQueueService,
            reachabilityRepository = reachabilityRepository,
            imageStorageRepository = imageStorageRepository
        )
    }

    val userProfileUseCase: UserProfileUseCase by lazy {
        UserProfileUseCaseImpl(
            userGateway = userGateway,
            userRepository = userRepository,
            authSessionRepository = authSessionRepository,
            syncQueueService = syncQueueService,
            reachabilityRepository = reachabilityRepository,
            imageStorageRepository = imageStorageRepository
        )
    }

    val syncQueuesUseCase: SyncQueuesUseCase by lazy {
        SyncQueuesUseCaseImpl(
            syncQueueRepository = syncQueueRepository,
            albumRepository = albumRepository,
            memoryRepository = memoryRepository,
            userRepository = userRepository
        )
    }
}
